dropshot/src/server.rs

```diff
@@ -327,6 +332,39 @@ impl<C: ServerContext> InnerHttpServerStarter<C> {
     }
 }
 
+/// This accepts connections more gracefully from a TcpListener.
+struct HttpAcceptor {
+    tcp: TcpListener,
+    log: slog::Logger,
+}
+
+impl HttpAcceptor {
+    async fn accept(&self) -> std::io::Result<(TcpStream, SocketAddr)> {
+        loop {
+            match self.tcp.accept().await {
+                Ok((socket, addr)) => return Ok((socket, addr)),
+                Err(e) => match e.kind() {
+                    // These are errors on the individual socket that we
+                    // tried to accept, and so can be ignored.
+                    std::io::ErrorKind::ConnectionRefused
+                    | std::io::ErrorKind::ConnectionAborted
+                    | std::io::ErrorKind::ConnectionReset => (),
+
+                    // This could EMFILE implying resource exhaustion.
+                    // Sleep a little bit and try again.
+                    _ => {
+                        warn!(self.log, "accept error"; "error" => e);
+                        tokio::time::sleep(std::time::Duration::from_millis(
+                            100,
+                        ))
+                        .await;
+                    }
```

Comment:
This is probably okay but I'm a little nervous about the error handling.  This is definitely an area where the standards and manual pages are good but operational experience is also relevant, which is why I'm still a little uneasy about implementing this behavior here instead of using a hardened library.  But I guess hyper 0.x was doing this inline before?  Is this the same logic it used?

---

The ECONNABORTED check is definitely right and it's one a lot of software misses so thanks for getting that. ✔️ 

I checked out the [illumos](https://illumos.org/man/3SOCKET/accept) manual page (our production systems), [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/functions/accept.html), and [Linux](https://www.man7.org/linux/man-pages/man2/accept.2.html).  I think most of the errors fall into one of these buckets:

* "shouldn't ever happen and if they did you probably don't want to accept any more": EBADF, EFAULT, ENODEV, ENOTSOCK, EOPNOTSUPP, EPROTO, EINVAL, EAGAIN/EWOULDBLOCK
* "can totally happen and should just be retried": ECONNABORTED, EINTR
* "probably shouldn't happen, but definitely can, and there's probably not a great response but we could retry, and maybe pausing would help": EMFILE, ENOMEM, ENOSR, ENOBUFS, EPERM? (Linux)

I don't see ECONNREFUSED or ECONNRESET as possible errors that can be produced here, but the Linux manual page does say any other error for the _new_ socket can happen here.  That seems weird to me, but okay.

It seems brittle to encode all of these, so I guess the logic here is saying: assume the "can't happen" stuff really can't happen, handle ECONNABORTED[/ECONNREFUSED/ECONNRESET] explicitly as transient errors that don't require pausing, and treat everything else as a transient error, with a pause to avoid getting stuck in too tight a loop if we're wrong about it being transient.  Is that the idea?

Reply:
Correct, this is essentially the same code that hyper 0.14 was doing (minus some TCP keep-alive options stuff, which Dropshot didn't enable): https://github.com/hyperium/hyper/blob/0.14.x/src/server/tcp.rs#L211-L254

I do think it'd be nice to have this in hyper-util, so server frameworks don't all need a copy of this. There's a couple design questions to make it sufficiently generic (listener type, timer type, how to log, etc), such that I don't expect it to be ready to be depended on immediately.